### ISSUE TITLE
DbFunctions: instance functions on entities

### ISSUE THREAD

**Issue by divega** [aspnet/EntityFrameworkCore#9811]
Consider the scenario originally described at https://github.com/aspnet/EntityFrameworkCore/issues/9213#issuecomment-324115344 (about why instance functions on `DbContext` can be compelling):

```csharp
db.Posts.Where(p => db.PostReadCount(p.Id) > 5);
```
It could actually be even more compelling to be able to do this:

``` csharp
db.Posts.Where(p => p.ReadCount() > 5);
```

This could work for instance methods, computed properties or even extension method defined over the entity type. 

It would be convenient for the translation to automatically map the "this" entity instance to the key of the entity when invoking the database implementation of the function.

**Comment by ajcvickers** [comment]
@pmiddleton Let us know if you're interested in working on this.

**Comment by pmiddleton** [comment]
Yes

**Comment by regul4rj0hn** [comment]
Please also consider an scenario like the following: I need to efficiently check some information that is encrypted by a symmetric key in the database, across several tables. 

It would be nice if  we could do something like `_dbContext.IsEmailAvailable(email)` given the mapping of scalar function in the DB to the context:
```
[DbFunction("IsEmailAvailable", "dbo")]
public static bool IsEmailAvailable(string Email) => throw new NotImplementedException();
``` 
You don't always need to execute the function over an specific Entity.

Thanks

**Comment by pmiddleton** [comment]
@regul4rj0hn - Is email bound to an entity in your example?  If so then you can already do that today.  

If email is just a string coming in from the application then you can't.  Support for use case is included in #11129 but that is on hold pending the query redesign in 3.0

**Comment by regul4rj0hn** [comment]
Yeah, it's coming from the app that's the issue. Currently calling the scalar function the old-fashioned way as a work around. I guess that refactor will have to wait then. Thanks for the help, @pmiddleton !
